ops/testing.py

```diff
@@ -1507,32 +1605,32 @@ def _storage_remove(self, storage_id: str):
         index = int(index)
         self._storage_list[name].pop(index, None)
 
-    def action_get(self):
-        raise NotImplementedError(self.action_get)
+    def action_get(self):  # type:ignore
+        raise NotImplementedError(self.action_get)  # type:ignore
 
-    def action_set(self, results):
-        raise NotImplementedError(self.action_set)
+    def action_set(self, results):  # type:ignore
+        raise NotImplementedError(self.action_set)  # type:ignore
 
-    def action_log(self, message):
-        raise NotImplementedError(self.action_log)
+    def action_log(self, message):  # type:ignore
+        raise NotImplementedError(self.action_log)  # type:ignore
 
-    def action_fail(self, message=''):
-        raise NotImplementedError(self.action_fail)
+    def action_fail(self, message=''):  # type:ignore
+        raise NotImplementedError(self.action_fail)  # type:ignore
 
-    def network_get(self, endpoint_name, relation_id=None):
-        raise NotImplementedError(self.network_get)
+    def network_get(self, endpoint_name, relation_id=None):  # type:ignore
+        raise NotImplementedError(self.network_get)  # type:ignore
 
-    def add_metrics(self, metrics, labels=None):
-        raise NotImplementedError(self.add_metrics)
+    def add_metrics(self, metrics, labels=None):  # type:ignore
+        raise NotImplementedError(self.add_metrics)  # type:ignore
```

Comment:
Um.. I was thinking: a function that is not implemented (and raises an exception if called) should not have a return type annotation at all -- lacking something like `-> WillRaiseIfCalled`.

As to the `Model._ModelBackend` methods not being annotated, I think that pyright is smart enough to figure out `-> None` if there is no return statement, and inherit the return type from `_run` if there is.

Reply:
I think that I'm okay w/ the `type:ignore` comments for the time being. None of them can be used at present, so we won't be introducing breaking changes in working code if we later implement these routines and introduce typing that matches the production back end.

Reply:
But we should at least add the annotations to the production backend methods (which are missing annotations) - that could possibly help eliminate the need for some type:ignores in various places.

Reply:
A missing type annotation (if not escaped with a # type:ignore) would make pyright complain. Pyright does not complain --> what more type annotations do you think we need?

Take add_metrics. `reveal_type(_ModelBackend.add_metrics` gives: `information: Type of "_ModelBackend.add_metrics" is "(self: _ModelBackend, metrics: Mapping[str, int | float], labels: Mapping[str, str] | None = None) -> None"`
which IMHO is all we need. Or are you saying we need more? Because if so, that's the case in **a lot** of places

Reply:
e.g. `storage_list` is missing a return type annotation (in model.py).  Basically the omitted type annotations here (which are okay since we raise exceptions) alerted me to the fact that we are missing some return type annotations in the production backend.

Reply:
and action_get - and maybe a few others.

Reply:
It's not necessarily related to this PR, I just saw it and thought it would be maybe worth addressing while you are hacking on types here.

Reply:
Ah now I get what you're saying. Even though pyright knows that it's a `List[int]` because it's a trivial inference, and therefore does not complain, we are still not 'documenting' the code by explicitly writing out the return type. 
Ok, I'm not against that, (although, what is the point? That's private code, and pyright knows that it's `List[int]`, so anyone who uses it will be warned if they misuse it) we can agree to do this for all 'hook tool' interface functions?
But why only those, then?

Reply:
I think it's a good habit - since it's not always obvious when pyright won't be able to figure it out.  This wasn't a hard suggestion/req - more like an "if you want to do this it might be nice".

Reply:
done. I also found that overloading yaml.safe_dump gets rid of a bunch of ignores.